#!/bin/bash
#PBS -k o
#PBS -l nodes=1:ppn=24,walltime=48:00:00,vmem=200GB
#PBS -M grthomas@indiana.edu
#PBS -m abe
#PBS -N om-angsd
#PBS -j oe

runCMD() {
    refdir="/N/u/grthomas/Carbonate/bin/referee"
    omdir="/N/dc2/scratch/grthomas/qtip/owl-monkey"
    omref="/N/dcwan/projects/hahnlab-phi/owl-monkey/ref-genome/genbank/GCA_000952055.2_Anan_2.0_genomic.fna"
    cd $omdir

    #angsd -GL 2 -i $omdir/split-bam-out/$1.bam -ref $omref -minQ 0 -doGlf 4 -remove_bads 0 -out om-angsd/$1
    # Run ANGSD on all scaffolds

    module load samtools/1.9
    samtools mpileup -d 999999999 -f $omref -Q 0 -B -x -o om-pileup/$1.pileup $omdir/split-bam-out/$1.bam
    # pileup command for no mapping quality (-s) and no baq recalculation (-B)
}

export -f runCMD
module load gnu-parallel/20170622
refdir="/N/u/grthomas/Carbonate/bin/referee"
omdir="/N/dc2/scratch/grthomas/qtip/owl-monkey"
omref="/N/dcwan/projects/hahnlab-phi/owl-monkey/ref-genome/genbank/GCA_000952055.2_Anan_2.0_genomic.fna"
cd $omdir
#cat scaffold-list.txt | parallel --eta 'runCMD'

cd $refdir
time -p referee.py -i $omdir/om-angsd.txt -ref $omref --correct --fastq -o $omdir/om-ref-angsd -p 24
# ANGSD inputs for Referee

#time -p referee.py -i $omdir/om-pileup.txt -ref $omref --pileup --correct --fastq -o $omdir/om-ref-pileup -p 24
# Pileup inputs for Referee

#module load r/3.4.4
#time -p Rscript helper-scripts/comp_scores.r $omdir/om-ref-angsd.txt $omdir/om-ref-pileup.txt $omdir/om-plots om-angsd-v-pileup
# Compare angsd vs pileup