---
title: "Example calculations"
---

<style type="text/css"> .table { width: 40%; } </style>

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
library(ggplot2)
library(directlabels)
library(gridExtra)
library(kableExtra)
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
source("lib/utils.r")
```



## I. Genotype likelihoods

Referee can accept genotype likelihoods calculated with any method as long as they are properly formatted. For Referee's internal genotype likelihood calculations we use the simple method [described in the initial GATK publication (McKenna et al 2010)](https://dx.doi.org/10.1101/gr.107524.110). What follows is a description of that method, with the additional consideration of mapping quality.

There are 10 possible genotypes at a biallelic site:

$$ \mathbb{G} = \{\{A,A\}, \{A,T\}, \{A,C\}, \{A,G\}, \{T,T\}, \{T,C\}, \{T,G\}, \{C,C\}, \{C,G\}, \{G,G\}\} $$

Given a set of reads $R$ at a site, we can calculate the likelihood of each genotype.

**Equation 1**

$$ P(R|\mathbb{G} = \{A_1,A_2\}) = \prod_{r}^{R} (\frac{1}{2}P(b_r|A_1) + \frac{1}{2}P(b_r|A_2) ) $$

To calculate the probability of the mapped base at the current read $b_r$ given each allele in the genotype $A_i$, we can use the Phred scaled base quality score $Q_b$ and Phred scaled mapping quality score $Q_m$ for that base. First we convert the Phred scale scores into probabilities of error:

**Equation 2**

$$ e = 10^{-\frac{Q}{10}} $$

Plugging in $Q_b$ and $Q_m$ gets us $e_b$ and $e_m$, respectively. These are the probabilities of error for both base calling and mapping.

Then we can calculate the probability of the base given an allele $P(b_r|A_i)$ as follows:

**Equation 3**

\begin{equation}
  P(b_r|A_i) =
    \begin{cases}
      \frac{(e_s * e_m)}{3} & \text{: $b \neq A_i$}\\
      1 - (e_s * e_m) & \text{: $b = A_i$}\\
    \end{cases}       
\end{equation}

Practically, the multiplication of many small probabilities in **Equation 1** can lead to underflow errors. To resolve this, we simply take the log of the probabilities and add them instead:

**Equation 4**

$$ P_{log}(R|\mathbb{G} = \{A_1,A_2\}) = \sum_{r}^{R} \log(\frac{1}{2}P(b_r|A_1) + \frac{1}{2}P(b_r|A_2)) $$

The log-likelihood of each genotype is obtained in this fashion. The log-likelihood for each of the 10 genotypes ($P_{log}(R|\mathbb{G} = \{A_1,A_2\})$) is all we need for the purpose of calculating a reference quality score. (Note: ANGSD does not consider mapping quality when calculating genotype likelihoods).

-----

## II. A reference quality score

First, we sum up the likelihoods of all genotypes that contain the reference allele ($L_{match}$) and separately sum up the likelihoods of all the genotypes that do not contain the reference allele ($L_{mismatch}$). 

**Equation 5**

$$ L_{match} = \sum_g^\mathbb{G} P(R\;|\;g) \; \text{if} \; b_R \in g $$

**Equation 6**

$$ L_{mismatch} = \sum_g^\mathbb{G} P(R\;|\;g) \; \text{if} \; b_R \notin g $$

For instance, if our reference base was an A, then:

$$ L_{match} = P(R|\{A,A\}) + P(R|\{A,T\}) + P(R|\{A,C\}) + P(R|\{A,G\})$$

and:

$$ L_{mismatch} = P(R|\{T,T\}) + P(R|\{T,C\}) + P(R|\{T,G\}) + P(R|\{C,C\}) + P(R|\{C,G\}) + P(R|\{G,G\})$$

with the probabilities being calculated with Equation 4.

We can then set up a likelihood ratio $LR$ by divding $L_{match}$ by $L_{mismatch}$:

**Equation 7**

$$ LR = \frac{L_{match}}{L_{mismatch}} $$

And this can be log-scaled to get us an informative value for a quality score:

**Equation 8**

$$ Q_\mathbb{R} = \log{LR} $$

## III. Calculation of $Q_\mathbb{R}$ on example read sets

**Example read sets:**

```{r read-sets, echo=FALSE}
read_set = c("A","A","A","A","A","A","A","A","A","A","A","A","A","A","A","A","A","A","A","A")
read_set_1a = list(num=1,reads=read_set,ref="A",ans="(correct)")
read_set_1t = list(num=1,reads=read_set,ref="T",ans="(incorrect)")
read_set_1c = list(num=1,reads=read_set,ref="C",ans="(incorrect)")
read_set = c("A","A","A","A","A","A","A","A","A","A","A","A","A","A","A","A","A","A","A","T")
read_set_2a = list(num=2,reads=read_set,ref="A",ans="(correct)")
read_set_2t = list(num=2,reads=read_set,ref="T",ans="(incorrect)")
read_set_2c = list(num=2,reads=read_set,ref="C",ans="(incorrect)")
read_set = c("A","A","A","A","A","A","A","A","A","A","T","T","T","T","T","T","T","T","T","T")
read_set_3a = list(num=3,reads=read_set,ref="A",ans="(correct)")
read_set_3t = list(num=3,reads=read_set,ref="T",ans="(correct)")
read_set_3c = list(num=3,reads=read_set,ref="C",ans="(incorrect)")

rs1_df = data.frame(read_set=c("Reads","Base qual","Map qual"),
                       reads=c("A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A",
                               "40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 1-40",
                               "40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 1-40"))

rs1_table = kable(rs1_df, "html", caption="Read set 1") %>%
  kable_styling(bootstrap_options=c("striped", "condensed", "responsive"), full_width=F)
gsub("<thead>.*</thead>", "", rs1_table)

rs2_df = data.frame(read_set=c("Reads","Base qual","Map qual"),
                       reads=c("A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  T",
                               "40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 1-40",
                               "40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 1-40"))

rs2_table = kable(rs2_df, "html", caption="Read set 2") %>%
  kable_styling(bootstrap_options=c("striped", "condensed", "responsive"), full_width=F)
gsub("<thead>.*</thead>", "", rs2_table)

rs3_df = data.frame(read_set=c("Reads","Base qual","Map qual"),
                       reads=c("A  A  A  A  A  A  A  A  A  A  T  T  T  T  T  T  T  T  T  T",
                               "40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 1-40",
                               "40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 1-40"))

rs3_table = kable(rs3_df, "html", caption="Read set 3") %>%
  kable_styling(bootstrap_options=c("striped", "condensed", "responsive"), full_width=F)
gsub("<thead>.*</thead>", "", rs3_table)
```

The above read sets were plugged into the relevant equations to calculate $Q_\mathbb{R}$ with one read varying in both base and mapping quality. The countour plots show how each behave with different reference base calls. 

In the figure below, each panel represents the results of genotype likelihood calculations and reference quality score calculations on a combination of one reference base and one read set with single read varying in base (x-axis) and mapping quality (y-axis). The shading indicates the most likely genotype, while the labeled dashed lines indicate the reference quality scores.

For example, in the upper left panel with read set 1 (all As) and a reference base call of A we see high scores regardless of varying quality of a single read. However, in the lower left panel the same read set when the reference base is called as a C scores very low, indicating that the reads do not support a C as the called base.


**Examples of $Q_{ref}$**

```{r method2, fig.width=14, fig.height=11.2, fig.align="center", echo=FALSE, message=FALSE}
method = 2
read_set_1a_results = cycleReads(read_set_1a, method)
rs1a_p = plotScores(read_set_1a_results, read_set_1a$num, read_set_1a$ref, read_set_1a$ans)
read_set_1t_results = cycleReads(read_set_1t, method)
rs1t_p = plotScores(read_set_1t_results, read_set_1t$num, read_set_1t$ref, read_set_1t$ans)
read_set_1c_results = cycleReads(read_set_1c, method)
rs1c_p = plotScores(read_set_1c_results, read_set_1c$num, read_set_1c$ref, read_set_1c$ans)
read_set_2a_results = cycleReads(read_set_2a, method)
rs2a_p = plotScores(read_set_2a_results, read_set_2a$num, read_set_2a$ref, read_set_2a$ans)
read_set_2t_results = cycleReads(read_set_2t, method)
rs2t_p = plotScores(read_set_2t_results, read_set_2t$num, read_set_2t$ref, read_set_2t$ans)
read_set_2c_results = cycleReads(read_set_2c, method)
rs2c_p = plotScores(read_set_2c_results, read_set_2c$num, read_set_2c$ref, read_set_2c$ans)
read_set_3a_results = cycleReads(read_set_3a, method)
rs3a_p = plotScores(read_set_3a_results, read_set_3a$num, read_set_3a$ref, read_set_3a$ans)
read_set_3t_results = cycleReads(read_set_3t, method)
rs3t_p = plotScores(read_set_3t_results, read_set_3t$num, read_set_3t$ref, read_set_3t$ans)
read_set_3c_results = cycleReads(read_set_3c, method)
rs3c_p = plotScores(read_set_3c_results, read_set_3c$num, read_set_3c$ref, read_set_3c$ans)

method2 = grid.arrange(rs1a_p,rs2a_p,rs3a_p,rs1t_p,rs2t_p,rs3t_p,rs1c_p,rs2c_p,rs3c_p, ncol=3,nrow=3)

result_list = list(read_set_1a_results, read_set_2a_results, read_set_3a_results,
                    read_set_1t_results, read_set_2t_results, read_set_3t_results,
                    read_set_1c_results, read_set_2c_results, read_set_3c_results)
#topScores(result_list, 40)
#topScores(result_list, 1)
```

This has the desired behavior of being a high score when we are sure the reference base is correct and a low score when we are sure the reference base is not correct. In fact, it has the nice property of being centered around 0, with positive scores indicating support for the called reference base and negative scores indicating support for the reference base being an error. The closer to 0 the score is (positive or negative) the less confident we are in our assertion.


