<!doctype html>
	<head>
		<!-- Global site tag (gtag.js) - Google Analytics -->
		<script async src="https://www.googletagmanager.com/gtag/js?id=UA-98263199-2"></script>
		<script>
			window.dataLayer = window.dataLayer || [];
			function gtag(){dataLayer.push(arguments);}
			gtag('js', new Date());

			gtag('config', 'UA-98263199-2');
        </script>
        
        <title>Referee calcs</title>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1">	
		<link type="text/css" rel="stylesheet" href="css/pure.css"  media="screen,projection" />
		<link type="text/css" rel="stylesheet" href="css/scaffold.css"  media="screen,projection" />
		<link rel='shortcut icon' href='img/favicon.png' type='image/x-icon'/ >
	</head>

<body>
	<div class="pure-g" id="nav_row">
		<div class="pure-u-3-24" id="margin"></div>
		<div class="pure-u-18-24" id="nav_col">
			<a href="#"><img class="pure_img" id="logo_nav" src="img/favicon.png"></a>
			<a class="title" href="index.html">Referee</a>
			<div class="pure-menu pure-menu-horizontal" id="nav_bar">
				<ul class="pure-menu-list" id="nav_list">
					<li class="pure-menu-item"><a href="index.html" class="nav_link">About</a></li>
                    <li class="pure-menu-item"><a href="#" class="nav_link" id="active">Calculations</a></li>
                    <li class="pure-menu-item"><a href="readme.html" class="nav_link">README</a></li>
					<li class="pure-menu-item"><a href="walkthrough.html" class="nav_link">Walkthrough</a></li>
					<li class="pure-menu-item"><a href="https://github.com/gwct/referee/releases/latest" target="_blank" class="nav_link">Download</a></li>
					<!-- <li class="pure-menu-item pure-menu-has-children pure-menu-allow-hover">
						<a href="#" id="menuLink1" class="pure-menu-link">Examples</a>
							<ul class="pure-menu-children">
								<li class="pure-menu-item"><a href="grampa/example1.html" class="pure-menu-link">Identify mode of polyploidy</a></li>
								<li class="pure-menu-item"><a href="grampa/example2.html" class="pure-menu-link">Place WGD on phylogeny</a></li>
								<li class="pure-menu-item"><a href="grampa/example3.html" class="pure-menu-link">Count duplications and losses</a></li>
							</ul>
					</li>
					<li class="pure-menu-item pure-menu-has-children pure-menu-allow-hover">
						<a href="#" id="menuLink1" class="pure-menu-link">Results</a>
							<ul class="pure-menu-children">
								<li class="pure-menu-item"><a href="grampa/results/yeast/yeast.html" class="pure-menu-link">Baker's Yeast</a></li>
								<li class="pure-menu-item"><a href="grampa/results/wheat/wheat_all.html" class="pure-menu-link">Bread Wheat</a></li>
								<li class="pure-menu-item"><a href="grampa/performance.html" class="pure-menu-link">Performance</a></li>
							</ul>
					</li> -->
					<li class="pure-menu-item"><a href="links.html" class="nav_link">Other links</a></li>
				</ul>

				<div class="mobile_dropdown" id="mobile_nav">
					<a href="#" id="nav_link_dropdown">Navigation</a>
					<div class="mobile_dropdown_links">
						<ul class="pure-menu-list">
							<li><a href="index.html" class="mobile_nav_link">About</a></li>
                            <li><a href="#" class="mobile_nav_link" id="active_drop">Calculations</a></li>
                            <li><a href="readme.html" class="mobile_nav_link">README</a></li>
							<li><a href="walkthrough.html" class="mobile_nav_link">Walkthrough</a></li>	
							<li><a href="https://github.com/gwct/referee/releases/latest" target="_blank" class="mobile_nav_link">Download</a></li>
							<li><a href="links.html" class="mobile_nav_link">Other links</a></li>
					</div>
				</div>				
			</div>
		</div>
		<div class="pure-u-3-24" id="margin"></div>
	</div>

	<div class="pure-g"><div class="pure-u-1" id="divider_row"></div></div>

	<div class="pure-g" id="main_row">
		<div class="pure-u-3-24" id="margin"></div>
		<div class="pure-u-18-24" id="main_col">
			<div id="main_content">
				<!-- <img class="pure-img" id="logo_main" src="img/ref-logo.png"> -->
				<h1>Referee's calculations</h1>
				<h3>This page is a summary of the calculations that Referee uses to obtain a reference quality score 
                    <span class="math inline">\(Q_\mathbb{R}\)</span> at a given site in a genome.</h3>

				<h2>I. Genotype likelihoods</h2>
                <p>Referee can accept genotype log-likelihoods calculated with any method as long as they are properly 
                    formatted (see <a href="readme.html">README</a>). For Referee's internal genotype likelihood 
                    calculations we use the simple method <a href="https://dx.doi.org/10.1101/gr.107524.110" target="_blank">described in the 
                    initial GATK publication (McKenna et al 2010)</a>. What follows is a description of that method, with the 
                    additional consideration of mapping quality.</p>

                <p>There are 10 possible genotypes at a biallelic site:</p>

                <p><span class="math display">\[ \mathbb{G} = \{\{A,A\}, \{A,T\}, \{A,C\}, \{A,G\}, \{T,T\}, \{T,C\}, \{T,G\}, \{C,C\}, \{C,G\}, \{G,G\}\} \]</span></p>
                
                <p>Given a set of reads <span class="math inline">\(R\)</span> at a site, we can calculate the likelihood of each genotype:</p>

                <p><strong>Equation 1A &#8212; diploid likelihood calculation</strong></p>
                
                <p><span class="math display">\[ P(R\;|\;g = \{A_1,A_2\}) = \prod_{r}^{R} (\frac{1}{2}P(b_r|A_1) + \frac{1}{2}P(b_r|A_2) ) \]</span></p>

                <p>We have also implemented a likelihood calculation for the most likely base in a set of reads from haploid data, in which the possible "genotypes" are:</p>

                <p><span class="math display">\[ \mathbb{G} = \{\{A\}, \{T\}, \{C\}, \{G\}\} \]</span></p>

                <p><strong>Equation 1B &#8212; haploid likelihood calculation</strong></p>
                
                <p><span class="math display">\[ P(R\;|\;g = \{A_1\}) = \prod_{r}^{R} (P(b_r|A_1)) \]</span></p>              

                <p>To calculate the probability of the mapped base at the current read <span class="math inline">\(b_r\)</span> given each allele in the 
                    genotype <span class="math inline">\(A_i\)</span>, we can use the Phred scaled base quality score <span class="math inline">\(Q_b\)</span> 
                    and Phred scaled mapping quality score <span class="math inline">\(Q_m\)</span> for that base. First we convert the Phred scaled 
                    scores into probabilities of error:</p>

                <p><strong>Equation 2 &#8212; Error probability from base quality</strong></p>
                <p><span class="math display">\[ e = 10^{-\frac{Q}{10}} \]</span></p>

                <p>Plugging in <span class="math inline">\(Q_b\)</span> and <span class="math inline">\(Q_m\)</span> gets 
                    us <span class="math inline">\(e_b\)</span> and <span class="math inline">\(e_m\)</span>, respectively. These are 
                    the probabilities of error for both base calling and mapping.</p>

                <p>Then we can calculate the probability of the base given an allele <span class="math inline">\(P(b_r|A_i)\)</span> as follows:</p>

                <p><strong>Equation 3 &#8212; The probability of a base given an allele</strong></p>
                <span class="math display">\[\begin{equation}
                  P(b_r|A_i) =
                    \begin{cases}
                      \frac{(e_b * e_m)}{3} &amp; \text{: $b \neq A_i$}\\
                      1 - (e_b * e_m) &amp; \text{: $b = A_i$}\\
                    \end{cases}       
                \end{equation}\]</span>

                <p>Practically, the multiplication of many small probabilities in <strong>Equation 1</strong> can lead to underflow errors. 
                    To resolve this, we simply take the log of the probabilities and add them instead:</p>

                <p><strong>Equation 4A &#8212; Log likelihood calculation for diploid data</strong></p>
                <p><span class="math display">\[ P_{log}(R\;|\;g = \{A_1,A_2\}) = \sum_{r}^{R} \log(\frac{1}{2}P(b_r|A_1) + \frac{1}{2}P(b_r|A_2)) \]</span></p>

                Similarly, for haploid data:
                <p><strong>Equation 4B &#8212; Log likelihood calculation for haploid data</strong></p>
                <p><span class="math display">\[ P_{log}(R\;|\;g = \{A_1\}) = \sum_{r}^{R} \log(P(b_r|A_1)) \]</span></p>

                <p>The log-likelihood of each genotype is obtained in this fashion. The log-likelihood for each of the 10 genotypes
                    <span class="math inline">\(P_{log}(R;|\;g = \{A_1,A_2\})\)</span> (or 4 bases for haploid data) is all we need for the purpose of calculating a reference 
                    quality score. (Note: ANGSD does not consider mapping quality when calculating genotype likelihoods).</p>

                <hr />

                <h2>II. A reference quality score</h2>
                <p>To summarize the quality information contained in the reads mapped to a site we first sum up the likelihoods of all genotypes that 
                    contain the reference allele (<span class="math inline">\(L_{match}\)</span>) and separately sum up the likelihoods of all 
                    the genotypes that do not contain the reference allele (<span class="math inline">\(L_{mismatch}\)</span>).</p>

                <p><strong>Equation 5</strong></p>
                <p><span class="math display">\[ L_{match} = \sum_g^\mathbb{G} P(R\;|\;g) \; \text{if} \; b_R \in g \]</span></p>

                <p><strong>Equation 6</strong></p>
                <p><span class="math display">\[ L_{mismatch} = \sum_g^\mathbb{G} P(R\;|\;g) \; \text{if} \; b_R \notin g \]</span></p>

                <p>For instance, if our reference base was an A, then:</p>

                <p><span class="math display">\[ L_{match} = P(R|\{A,A\}) + P(R|\{A,T\}) + P(R|\{A,C\}) + P(R|\{A,G\})\]</span></p>

                <p>and:</p>

                <p><span class="math display">\[ L_{mismatch} = P(R|\{T,T\}) + P(R|\{T,C\}) + P(R|\{T,G\}) + P(R|\{C,C\})\]</span></p>
                <p><span class="math display">\[ + P(R|\{C,G\}) + P(R|\{G,G\})\]</span></p>

                <p>with the probabilities being calculated with <strong>Equation 4</strong>.</p>

                <p>We then set up a likelihood ratio <span class="math inline">\(LR\)</span> by divding <span class="math inline">\(L_{match}\)</span> 
                    by <span class="math inline">\(L_{mismatch}\)</span>:</p>

                <p><strong>Equation 7</strong></p>
                <p><span class="math display">\[ LR = \frac{L_{match}}{L_{mismatch}} \]</span></p>

                <p>And this can be log-scaled to get us an informative value for a quality score:</p>

                <p><strong>Equation 8</strong></p>
                <p><span class="math display">\[ Q_\mathbb{R} = \log{LR} \]</span></p>
                
                <hr />

                <h2>III. Calculation of <span class="math inline">\(Q_\mathbb{R}\)</span> on example read sets</h2>
                <p><strong>Example read sets:</strong></p>

                <table class="pure-table pure-table-striped" style="width: auto !important; margin-left: auto; margin-right: auto;">
                    <caption>Read set 1</caption>
                    <tbody>
                    <tr>
                        <td style="text-align:left;">Reads</td>
                        <td style="text-align:left;">A A A A A A A A A A A A A A A A A A A A</td>
                    </tr>
                    <tr>
                        <td style="text-align:left;">Base quals</td>
                        <td style="text-align:left;">40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 1-40</td>
                    </tr>
                    <tr>
                        <td style="text-align:left;">Map quals</td>
                        <td style="text-align:left;">40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 1-40</td>
                    </tr>
                    </tbody>
                </table>
                </br>
                <table class="pure-table pure-table-striped" style="width: auto !important; margin-left: auto; margin-right: auto;">
                    <caption>Read set 2</caption>
                    <tbody>
                    <tr>
                        <td style="text-align:left;">Reads</td>
                        <td style="text-align:left;">A A A A A A A A A A A A A A A A A A A T</td>
                    </tr>
                    <tr>
                        <td style="text-align:left;">Base quals</td>
                        <td style="text-align:left;">40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 1-40</td>
                    </tr>
                    <tr>
                        <td style="text-align:left;">Map quals</td>
                        <td style="text-align:left;">40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 1-40</td>
                    </tr>
                    </tbody>
                </table>                
                </br>
                <table class="pure-table pure-table-striped" style="width: auto !important; margin-left: auto; margin-right: auto;">
                    <caption>Read set 3</caption>
                    <tbody>
                    <tr>
                        <td style="text-align:left;">Reads</td>
                        <td style="text-align:left;">A A A A A A A A A A T T T T T T T T T T</td>
                    </tr>
                    <tr>
                        <td style="text-align:left;">Base quals</td>
                        <td style="text-align:left;">40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 1-40</td>
                    </tr>
                    <tr>
                        <td style="text-align:left;">Map quals</td>
                        <td style="text-align:left;">40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 40 1-40</td>
                    </tr>
                    </tbody>
                </table>


                <p>The above read sets were plugged into the relevant equations to calculate <span class="math inline">\(Q_\mathbb{R}\)</span> 
                    with one read varying in both base and mapping quality. The countour plots show how each behave with different reference base calls.</p>
                <p>In the figure below, each panel represents the results of genotype likelihood calculations and reference quality score 
                    calculations on a combination of one reference base and one read set with single read varying in base (x-axis) and mapping 
                    quality (y-axis). The shading indicates the most likely genotype, while the labeled dashed lines indicate the reference quality scores.</p>
                <p>For example, in the upper left panel with read set 1 (all As) and a reference base call of A we see high scores regardless of 
                    varying quality of a single read. However, in the lower left panel the same read set when the reference base is called as a C 
                    scores are very low, indicating that the reads do not support a C as the called base.</p>
               
                <p><strong>Examples of <span class="math inline">\(Q_{ref}\)</span></strong></p>
                <p><img src="img/method2-1.png" class="pure_img" id="contour_plot" /></p>

                <p>This has the desired behavior of being a high score when we are sure the reference base is correct and a low score when we are 
                    sure the reference base is not correct. In fact, it has the nice property of being centered around 0, with positive scores 
                    indicating support for the called reference base and negative scores indicating support for the reference base being an error. 
                    The closer to 0 the score is (positive or negative) the less confident we are in our assertion.</p>
            </div>
		</div>
		<div class="pure-u-3-24" id="margin"></div>
	</div>

	<div class="pure-g"><div class="pure-u-1" id="divider_row"></div></div>
	<div class="pure-g" id="footer">
		<div class="pure-u-1">
			<div id="footer_text">
				<center>Site designed and maintained by <a href="https://gwct.github.io/index.html" target="_blank">Gregg Thomas</a> | Some of the CSS used to design
					this site is from the <a href="https://purecss.io/" target="_blank">Pure CSS</a> project.</center>
			</div>
		</div>
    </div>
    
    <!-- dynamically load mathjax for compatibility with self-contained -->
    <script>
        (function () {
          var script = document.createElement("script");
          script.type = "text/javascript";
          script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
          document.getElementsByTagName("head")[0].appendChild(script);
        })();
    </script>

</body>
</html>