---
title: "Referee"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
</br>

<center> ![](img/ref-icon-4.png) </center>
</br>

### Referee is a program to calculate a quality score for every position in a genome assembly. This allows for easy filtering of low quality sites for any downstream analysis.

Modern genome sequencing technologies provide a succint measure of quality at each position in every read, however all of this information is lost in the assembly process. Referee summarizes the quality information from the reads that map to a site in an assembled genome to calculate a quality score for each position in the genome assembly. We accomplish this by first calculating genotype likelihoods for every site. For a given site in a diploid genome, there are 10 possible genotypes (AA, AC, AG, AT, CC, CG, CT, GG, GT, TT). Referee takes as input the genotype likelihoods calculated for all 10 genotypes given the called reference base at each position. To obtain these likelihoods, one must first map the reads used to make the assembly back onto the finished assembly. Then these reads can be used to calculate genotype likelihoods using any method/program. [Referee can calculate](example.html) the likelihoods from a pileup file as input or use pre-calculated log likelihoods, such as those output by [ANGSD](https://github.com/ANGSD/angsd). Then, Referee compares the log of the ratio of the sum of genotype likelihoods for genotypes that contain the reference base vs. the sum of those that do not contain the reference base. Positive scores indicate support for the called reference base while negative scores indicate support for some other base. Scores close to 0 indicate less confidence while higher scores indicate more confidence in the reference base. Scores range from 0 to 91, with some special cases (see [README](https://github.com/gwct/referee)). With the `--correct` option specified Referee will also output the highest scoring base for sites with negative scores.

## Usage

These are the general steps for scoring your genome:

1. Using any applicable software, map the reads from which you constructed your genome back to the finished assembly. (A BAM file is usable by ANGSD for calculating genotype likelihoods in the next step)

2. Compile a pileup file for Referee to calculate genotype likelihoods OR pre-calculate log genotype likelihoods for all 10 genotypes at every position in the genome (we recommend [ANGSD](https://github.com/ANGSD/angsd) for this).

3. Score your genome with one of the following Referee commands:

`python referee.py -gl [genotype likelihood file] -ref [reference genome FASTA file] --pileup`

Alternatively, if you have multiple genotype likelihood files you wish to score with the same reference genome, you could put the paths to each file in a text file with one file path per line for Referee to score them all:

`python referee.py -i [text file with paths to genotype likelihood files] -ref [reference genome FASTA file] --pileup`

If you have pre-calculated genotype likelihoods as input, exclude the `--pileup` flag.

</br>
<center>
### **For more information on the usage and inputs, see the [README](https://github.com/gwct/referee)**
</center>