<!doctype html>
	<head>
		<title>Performance</title>
		<link type="text/css" rel="stylesheet" href="css/pure.css"  media="screen,projection" />
		<link type="text/css" rel="stylesheet" href="css/scaffold.css"  media="screen,projection" />
		<link rel='shortcut icon' href='img/favicon.png' type='image/x-icon'/ >
	</head>

<body>
	<div class="pure-g" id="nav_row">
		<div class="pure-u-1-5" id="margin"></div>
		<div class="pure-u-3-5" id="nav_col">
			<a href="#"><img class="pure_img" id="logo_nav" src="img/favicon.png"></a>
			<a class="title" href="#">Referee</a>
			<div class="pure-menu pure-menu-horizontal" id="nav_bar">
				<ul class="pure-menu-list" id="nav_list">
					<li class="pure-menu-item"><a href="index.html" class="nav_link" id="active">About</a></li>
					<li class="pure-menu-item"><a href="calcs.html" class="nav_link">Calculations</a></li>
					<li class="pure-menu-item"><a href="readme.html" class="nav_link">README</a></li>
					<li class="pure-menu-item"><a href="walkthrough.html" class="nav_link">Walkthrough</a></li>
					<li class="pure-menu-item"><a href="https://github.com/gwct/referee" target="_blank" class="nav_link">Download</a></li>
					<!-- <li class="pure-menu-item pure-menu-has-children pure-menu-allow-hover">
						<a href="#" id="menuLink1" class="pure-menu-link">Examples</a>
							<ul class="pure-menu-children">
								<li class="pure-menu-item"><a href="grampa/example1.html" class="pure-menu-link">Identify mode of polyploidy</a></li>
								<li class="pure-menu-item"><a href="grampa/example2.html" class="pure-menu-link">Place WGD on phylogeny</a></li>
								<li class="pure-menu-item"><a href="grampa/example3.html" class="pure-menu-link">Count duplications and losses</a></li>
							</ul>
					</li>
					<li class="pure-menu-item pure-menu-has-children pure-menu-allow-hover">
						<a href="#" id="menuLink1" class="pure-menu-link">Results</a>
							<ul class="pure-menu-children">
								<li class="pure-menu-item"><a href="grampa/results/yeast/yeast.html" class="pure-menu-link">Baker's Yeast</a></li>
								<li class="pure-menu-item"><a href="grampa/results/wheat/wheat_all.html" class="pure-menu-link">Bread Wheat</a></li>
								<li class="pure-menu-item"><a href="grampa/performance.html" class="pure-menu-link">Performance</a></li>
							</ul>
					</li> -->
					<li class="pure-menu-item"><a href="links.html" class="nav_link">Other links</a></li>
				</ul>

				<div class="mobile_dropdown" id="mobile_nav">
					<a href="#" id="nav_link_dropdown">Navigation</a>
					<div class="mobile_dropdown_links">
						<ul class="pure-menu-list">
							<li><a href="index.html" class="mobile_nav_link" id="active_drop">About</a></li>
							<li><a href="calcs.html" class="mobile_nav_link">Calculations</a></li>
							<li><a href="readme.html" class="mobile_nav_link">README</a></li>
							<li><a href="walkthrough.html" class="mobile_nav_link">Walkthrough</a></li>	
							<li><a href="https://github.com/gwct/referee" target="_blank" class="mobile_nav_link">Download</a></li>
							<li><a href="links.html" class="mobile_nav_link">Other links</a></li>
					</div>
				</div>				
			</div>
		</div>
		<div class="pure-u-1-5" id="margin"></div>
	</div>

	<div class="pure-g"><div class="pure-u-1" id="divider_row"></div></div>

	<div class="pure-g" id="main_row">
		<div class="pure-u-3-24" id="margin"></div>
		<div class="pure-u-18-24" id="main_col">
			<div id="main_content">
                <h1>Referee perfomance tests</h1>
                
                <h2><em>J. sinuosa</em> transcriptome</h2>
                    <p>The <em>J. sinuosa</em> reads for vegetative tissues were downloaded with fastq-dump from NCBI 
                        <a href="https://www.ncbi.nlm.nih.gov/sra/SRX2676125%5Baccn%5D">SRA SRX2676125</a>. These reads were 
                        assembled with Trinity using default parameters:</p>
                    <pre><code>Trinity --seqType fq --max_memory 100G  --left SRR5380902_1.fastq  --right SRR5380902_2.fastq --CPU 8</code></pre>

                    <p>The reads were mapped back to the resulting transcriptome assembly:</p>
                    <pre><code>bwa mem Trinity.fasta SRR5380902_1.fastq SRR5380902_2.fastq &gt; SRR5380902.sam</code></pre>

                    <p>A few additional steps were performed: 1) create the sequence dictionary to add to the SAM header because this output 
                        lacked the <span class="citation">@HD</span> line needed later on, 2) Add the header from the dict, 3) Convert to BAM, 4) Sort BAM.</p>

                    <p>Next, genotype likelihoods were calculated with ANGSD:</p>
                    <pre><code>angsd -GL 2 -i SRR5380902-sorted.bam -ref Trinity.fasta -minQ 0 -doGlf 4 -out jsinuosa-angsd</code></pre>

                    <p>And reference quality scores calculated with Referee:</p>
                    <pre><code>referee.py -gl jsinuosa-angsd.glf.gz -ref Trinity.fasta --fastq --correct -o jsinuosa-referee</code></pre>

                    <p>Additionally, two pileup files were created from the BAM file to test Refereeâ€™s genotype likelihood calculations. One with mapping quality and one without (the -s flag controls mapping quality output):</p>
                    <pre><code>samtools mpileup -d 999999999 -f &lt;reference.fa&gt; -Q 0 -s -o jsinuosa-mq.pileup SRR5380902-sorted.bam</code></pre>

                    <p>The pileup was used by Referee to calculate genotype likelihoods and quality scores:</p>
                    <pre><code>referee.py -gl jsinuosa-mq.pileup -ref Trinity.fasta --pileup --fastq --correct -o jsinuosa-referee-glcalcs-mq</code></pre>

                    <p>Here are some statistics from this process:</p>
                    <center><table class="pure-table pure-table-bordered pure-table-striped">
                        <thead><th>Description</th><th>Value</th></thead>
                        <tbody>
                            <tr><td>File size of assembled transcriptome</td> <td>53.96 MB</td></tr>
                            <tr><td># transcripts</td> <td>55,567</td></tr>
                            <tr><td># bases</td> <td>50,283,993</td></tr>
                            <tr><td>Size of sorted BAM file</td> <td>1.01 GB</td></tr>
                            <tr><td>Average read depth</td> <td>19.53</td></tr>
                            <tr><td>Time to run ANGSD</td> <td>7.98 minutes</td></tr>
                            <tr><td>Sites with ANGSD calcs</td> <td>48,791,504</td></tr>
                            <tr><td>Time to run Referee (1 proc)</td> <td>18.69 minutes</td></tr>
                            <tr><td>Size of Referee output (tab)</td> <td>1.67 GB</td></tr>
                            <tr><td>Size of Referee output (fastq)</td> <td>124.97 MB</td></tr>
                        </tbody>
                    </table></center>

                    <h3>Referee run time on <em>J. sinuosa</em> data</h3>
                        <p><strong>Figure 1: Referee runtime on J. sinuosa data with 3 different genotype likelihood methods</strong></p>
                        <center>
                        <img src="img/performance-plots/jsin-time.png" style="width:75.0%" />
                        </center>
                        <p></br></p>


                    <h3>Referee memory usage on <em>J. sinuosa</em> data</h3>
                        <p><strong>Figure 2: Referee max memory use on J. sinuosa data with 3 different genotype likelihood methods</strong></p>
                        <center>
                        <img src="img/performance-plots/jsin-mem.png" style="width:75.0%" />
                        </center>
                        <p></br></p>


                <h3>Comparing scores from different genotype likelihood methods on <em>J. sinuosa</em> data</h3>
                    <center><table class="pure-table pure-table-bordered pure-table-striped">
                        <thead>
                            <th>Calculation method</th>
                            <th><span class="math inline">\(L_{mismatch}\)</span> = 0 (<span class="math inline">\(R_Q\)</span> = 91)</th>
                            <th>Reference is N (<span class="math inline">\(R_Q\)</span> = -1)</th>
                            <th>No reads mapped (<span class="math inline">\(R_Q\)</span> = -2)</th>
                        </thead>
                        <tbody>
                            <tr><td>ANGSD</td> <td>1646845</td> <td>0</td> <td>1487227</td></tr>
                            <tr><td>Pileup without MQ</td> <td>1648828</td> <td>0</td> <td>1478554</td></tr>
                            <tr><td>Pileup with MQ</td> <td>4113594</td> <td>0</td> <td>1478554</td></tr>
                        </tbody>
                    </table></center>


                <h3>Referee score distributions and read depth correlations</h3>
                    <p><strong>Figure 3: Referee score 1 distribution on J. sinuosa</strong></p>
                    <center>
                    <img src="img/jsin-plots/jsin-ref-angsd-s1-score-hist.png" style="width:75.0%" />
                    </center>
                    <p></br></p>

                    <p><strong>Figure 4: Referee score 1 vs. read depth</strong></p>
                    <center>
                    <img src="img/jsin-plots/jsin-ref-angsd-s1-score-v-depth.png" style="width:75.0%" />
                    </center>
                    <p></br></p>

                    <p><strong>Figure 5: ANGSD score 1 vs. Pileup score 1</strong></p>
                    <center>
                    <img src="img/jsin-plots/jsin-angsds1-vs-pileups1-score-v-score.png" style="width:75.0%" />
                    </center>
                    <p></br></p>

                    <p><strong>Figure 5: ANGSD score 1 vs. Pileup score 2</strong></p>
                    <center>
                    <img src="img/jsin-plots/jsin-angsds1-vs-angsds2-score-v-score.png" style="width:75.0%" />
                    </center>
                    <p></br></p>

                    <p><strong>Figure 6: Pileup score 1 vs. Pileup score 2</strong></p>
                    <center>
                    <img src="img/jsin-plots/jsin-pileups1-vs-pileups2-score-v-score.png" style="width:75.0%" />
                    </center>
                    <p></br></p>



                <h3>Using mapping quality</h3>
                    <p><strong>Figure 7: Referee score 1 distribution on J. sinuosa with mapping quality</strong></p>
                    <center>
                    <img src="img/jsin-plots/jsin-ref-pileup-s1-mq-score-hist.png" style="width:75.0%" />
                    </center>
                    <p></br></p>

                    <p><strong>Figure 8: Referee score 1 with mapping quality vs. read depth</strong></p>
                    <center>
                    <img src="img/jsin-plots/jsin-ref-pileup-s1-mq-score-v-depth.png" style="width:75.0%" />
                    </center>
                    <p></br></p>

                    <p><strong>Figure 9: Pileup score 1 NO mapping quality vs. Pileup score 1 WITH mapping quality</strong></p>
                    <center>
                    <img src="img/jsin-plots/jsin-pileups1-vs-pileups1-mq-score-v-score.png" style="width:75.0%" />
                    </center>
                    <p></br></p>

            <h2>Owl monkey genome</h2>
                <p>The owl monkey reference genome was obtained from NCBI 
                    <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/952/055/GCA_000952055.2_Anan_2.0">GCA_000952055.2_Anan_2.0</a> and the 
                    reads were mapped back to this assembly by colleagues at Baylor for the Owl monkey pedigree project. The BAM file was split by 
                    scaffold for parallelizaiton of ANGSD calcs. This splitting shouldnâ€™t effect the Referee performance, though it means I donâ€™t 
                    have genome-wide score distributions.</p>

                <p>Here are some statistics from this genome:</p>
                <center><table class="pure-table pure-table-bordered pure-table-striped">
                    <thead><th>Description</th><th>Value</th></thead>
                    <tbody>
                        <tr><td>File size of assembled genome</td> <td>2.90 GB</td></tr>
                        <tr><td># scaffolds</td> <td>22,922</td></tr>
                        <tr><td># bases</td> <td>2,861,668,348</td></tr>
                        <tr><td>Size of sorted BAM file</td> <td>561.67 GB</td></tr>
                        <tr><td>Average read depth</td> <td>152.988</td></tr>
                        <tr><td>Time to run ANGSD</td> <td>54.38 minutes</td></tr>
                        <tr><td>Sites with ANGSD calcs</td> <td>2,664,619,299</td></tr>
                        <tr><td>Time to run Referee (1 proc)</td> <td>17.33 hours</td></tr>
                        <tr><td>Size of Referee output (tab)</td> <td>7.6 GB</td></tr>
                        <tr><td>Size of Referee output (fastq)</td> <td>608 MB</td></tr>
                        <tr><td><span class="math inline">\(L_{mismatch}\)</span> = 0 (<span class="math inline">\(R_Q\)</span> = 91)</td> <td>2,385,493,538</td></tr>
                        <tr><td>Reference is N (<span class="math inline">\(R_Q\)</span> = -1)</td> <td>173,395</td></tr>
                        <tr><td>No reads mapped (<span class="math inline">\(R_Q\)</</td> <td>197,049,049</td></tr>
                    </tbody>
                </table></center>

                    <h3>Referee run time on Owl monkey and <em>J. sinuosa</em> data</h3>
                        <p><strong>Figure 12: Referee runtime on J. sinuosa and owl monkey data with 3 different genotype likelihood methods</strong></p>
                        <center>
                        <img src="img/performance-plots/time.png" style="width:75.0%" />
                        </center>
                        <p></br></p>

                    <h3>Referee memory usage on Owl monkey and <em>J. sinuosa</em> data</h3>
                        <p><strong>Figure 13: Referee max memory use on J. sinuosa and owl monkey data with 3 different genotype likelihood methods</strong></p>
                        <center>
                        <img src="img/performance-plots/mem.png" style="width:75.0%" />
                        </center>
			</div>
		</div>
		<div class="pure-u-3-24" id="margin"></div>
	</div>

	<div class="pure-g"><div class="pure-u-1" id="divider_row"></div></div>
	<div class="pure-g" id="footer">
		<div class="pure-u-1">
			<div id="footer_text">
				<center>Site designed and maintained by <a href="https://gwct.github.io/index.html" target="_blank">Gregg Thomas</a> | Some of the CSS used to design
					this site is from the <a href="https://purecss.io/" target="_blank">Pure CSS</a> project.</center>
			</div>
		</div>
    </div>

    <!-- dynamically load mathjax for compatibility with self-contained -->
    <script>
        (function () {
            var script = document.createElement("script");
            script.type = "text/javascript";
            script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
            document.getElementsByTagName("head")[0].appendChild(script);
        })();
    </script>

</body>
</html>